Implementation of a API Mocking Server and testcases for the VCSA-Exporter

voigtst opened the pull request:
WIP PR. This PR shows the current progress on the GitHub Actions integration. The Flask Mocking Server is feature complete. The MockServer class contains the logic to setup the server. The method "start_server" spins up a WSGI Server. The classes Authentication, LoggingService, VmonService contain the logic that the Server needs. The RequestHandler class contains the central control logic for the other classes.
The testcases are WIP. With the WSGI it is possible to spin up a non blocking Flask Server for pytest (deamon thread). An example is attached to this PR.

voigtst commented:
Mocking Server and testcases are implemented.

Currently 5 tests are failing:

in test_connection:
test_login_logout_failure:
fails, because logout just logs it's status and therefore returns 'None' (2 fails)

test_logout_success:
fails, because logout returns 'None' instead of 'False' (1 fail)


in test_metric_output:
test_vcenter_backend_down:
because the vcsa currently appends metrics data during runtime, the backend down test fails. (1 fail)

test_consecutive_runs:
because the vcsa currently appends metrics data during runtime, the backend down test fails. (1 fail)

the GitHub Actions file still needs to be implemented. This will happen the following days.

viennaa commented on tests/test_json_output.py:
```diff
@@ -0,0 +1,51 @@
+
+
+class TestJsonOuput:
+    def test_yaml_read(self, bc_yaml):
```
simple and good, I like it.

voigtst replied:
Thank you very much!

voigtst replied:
I implemented the BaseCollector read in the test_yaml_read now. The fixture should read the data from the file. It then behaves like the other fixtures used in test_json_output.

viennaa commented on tests/test_json_output.py:
```diff
@@ -0,0 +1,51 @@
+
+
+@pytest.fixture(scope='class')
+def bc_yaml():
```
```suggestion
def rest_yaml():
```
would go as close as possible from the naming

voigtst replied:
good Idea.

viennaa commented on tests/test_json_output.py:
```diff
@@ -0,0 +1,51 @@
+        assert bc_yaml == rest_yaml
+
+    def test_connection_get(self, setup_vcenter, bc_yaml, vmon_json):
+        key = bc_yaml['vmonservice']['api_target']
```
```suggestion
        api_target = bc_yaml['vmonservice']['api_target']
```
Why key?

voigtst replied:
I will add this in a separate commit, because if I commit it here, I break the lines below. Key was chosen, because it mirrors the naming convention in the Connection class. But api_target is better to understand in this context.

voigtst replied:
done.

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+        response = requests.get(setup_vcsa_url)
+        metrics = self.process_response_data(response)
+
+        with open('tests/vcsa_output_comparison.yaml', 'r') as fp:
```
Would put into data dir where you have put the other 'static' files already

voigtst replied:
Will add the file to mockingServer/data

voigtst replied:
done

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+
+        for key in comparison_data.keys():
+            difference = set(comparison_data[key]).difference(metrics)
+            assert len(difference) == 0
```
You are expecting a difference here, which makes sense from what you are doing earlier. Although, what is the purpose of this test? A proper name would be `test_prometheus_server_reply` or comparable. You are just comparing the plain output of prometheus with an always divergent yaml.

voigtst replied:
I am not expecting a difference at this part. I compare the vcsa metrics with a static metric collection that I dumped from the vcsa at a working state.  If the output from the exporter differs from the expected yaml file (len(difference) > 0), the testcase fails. It is the main exporter test, which tests if the exporter as a whole is delivering the correct output.

voigtst replied:
done.

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+            assert len(difference) == 0
+
+    def test_vcenter_backend_down(self, request, setup_vcsa_url, setup_vcenter):
+        setup_vcenter.name = 'false'
```
How about
```suggestion
        setup_vcenter.name = 'foobar-vcenter.whatever.domain'
```

voigtst replied:
This is a good idea, it enhances readability.

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+            difference = set(comparison_data[key]).difference(metrics)
+            assert len(difference) == 0
+
+    def test_vcenter_backend_down(self, request, setup_vcsa_url, setup_vcenter):
```
This test is not testing how the exporter is interacting against a down backend. setup_vcenter would need another host here and then would need to do the query, wouldn't it?

voigtst replied:
The connection is done with the vcenter name, which gets assigned to the connection target attribute within the constructor of the Connection class. So it does sent a request against a non existing host. The VCSA exporter then returns default metrics created by Prometheus 
Could you elaborate a little bit further?

viennaa replied:
Alright, so it is supposed to just check for the empty prometheus server.
Interesting test case then would be to basically run
* `test_vcsa_data_comparsion`
* set the backend down 
* check ouput
* set backend up
* check output

voigtst replied:
I presume with the point `test_vcsa_data_comparsion` you mean the executed test before test_vcenter_backend_down?
If yes, I would implement it as you said, set backend down, check the output, set the backend up, check the output.

voigtst replied:
I implemented another test for that, called 'test_vcenter_backend_up'. This test checks if expected metrics are returning. The difference to 'test_prometheus_server_reply' is, that it checks if the returning data is correct.

viennaa replied:
That's a valid test case. Can we also add it in one test as described above? 
This then would be more like a real world scenario like backend is up, then it goes down, then it goes up again. Idea is to see, if the code can survive.

voigtst replied:
Refactored those testcase into 'test_backend_recovery_from_down'

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+        setup_vcenter.name = request.config.getoption('--host')
+        assert len(metrics) == 0
+
+    def test_consecutive_runs(self, setup_vcsa_url):
```
👍

voigtst replied:
I'm glad you like it :)

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,39 @@
+        metrics = set()
+
+        for entry in results:
+            if entry.startswith('vcsa_service_status') or entry.startswith('vcsa_logging_status'):
```
```suggestion
            if entry.startswith(('vcsa_service_status', 'vcsa_logging_status')):
```

voigtst replied:
Very elegant change.

viennaa commented on mockingServer/modules/RequestHandler.py:
```diff
@@ -0,0 +1,14 @@
+class RequestHandler(MethodView):
+    def __init__(self):
+        with open('mockingServer/data/login.json', 'r') as data:
+            self.session_id = json.load(data)
```
Performance is not really an issue here, but it does open this file every time as we are issuing requests, which is quite often. I would be finde with issuing a static value from here, we are having it in a file, which is not changing anyway, so it has no real benefit.

voigtst replied:
True, I will hardcode it and remove the login.json from the data subdirectory.

voigtst replied:
done

viennaa commented on mockingServer/modules/LoggingSession.py:
```diff
@@ -0,0 +1,17 @@
+class LoggingService(RequestHandler):
+    def post(self):
+        if not self.check_session_id():
+            return make_response('sessionID check failed', 401)
```
Is this the real response?

voigtst replied:
Yes, if you sent a request the message will be stored as a content attribute.
f.e: 
response = requests.post(...)
print(response.content) -> would provide this message.

viennaa replied:
Funny, ok

viennaa commented on mockingServer/modules/LoggingSession.py:
```diff
@@ -0,0 +1,17 @@
+    def post(self):
+        if not self.check_session_id():
+            return make_response('sessionID check failed', 401)
+        with open('mockingServer/data/logging.json') as data:
```
either use `r` everywhere for opening read only or skip it, you have mixed it.

voigtst replied:
Will add 'r' for better readability.

voigtst replied:
done

viennaa reviewed (changes requested):
A very solid first PR. I really like the Mockserver and Flask implementation and how it welds with pytest. Please find my questions and suggestions inline.

viennaa commented:
A more general note, consider using `self.assertEqual`, `self.assertFalse` etc to add error messages what is currently failing and why we expect that. For instance this is supposed to fail, because a login is missing: https://github.com/sapcc/vcsa-exporter/pull/7/files#diff-dfbf691ce15248c24632831234785fc221b10763ff99e326336d4ca328daa9d6R48
which is not really obvious if you don't have insight.

voigtst commented on mockingServer/MockServer.py:
```diff
@@ -0,0 +1,26 @@
+                                 certfile='mockingServer/certs/cert.cert')
+        self.server.serve_forever()
+
+    def stop_server(self):
+        self.server.stop()
```
Correct. Fixing that.

voigtst replied:
removed that part.

voigtst commented on mockingServer/MockServer.py:
```diff
@@ -0,0 +1,26 @@
+        self.app.register_blueprint(loggingBluePrint, url_prefix='/rest')
+
+    def start_server(self):
+        self.server = WSGIServer(('localhost', 443), self.app,
```
Will add self.port

voigtst replied:
added self.port as a parameter

voigtst commented on mockingServer/data/login.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "value": "MockServerSessionID"
```
I generated one. What do you think of "779b8ede-8332-11eb-9581-3c58c27e75a6"?

viennaa replied:
Well, how about some `1337` in there?

voigtst replied:
done.

voigtst commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+
+@pytest.fixture(scope='session')
+def setup_vcsa_url(request):
+    """To reduce maintenance we forge a VCSA URL for the tests."""
```
I will specify that to: 
"To avoid changing the URL throughout the testcases on changing parameters, we create a VCSA URL upfront."

voigtst commented on mockingServer/MockServer.py:
```diff
@@ -0,0 +1,26 @@
+class MockServer:
+    def __init__(self, port=443):
+        self.app = Flask(__name__)
+        self.port = port
```
It is now used as parameter to create the WSGI Server.

voigtst commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+
+    port = request.config.getoption('--prometheusport')
+    vcenter_list = [setup_vcenter]
+    thread = Thread(target=run_prometheus_server, args=(port, vcenter_list))
```
done

voigtst commented on tests/test_connection.py:
```diff
@@ -0,0 +1,34 @@
+import requests
+
+
+@pytest.fixture(scope='class')
+def session_id():
+    mock_session_id = RequestHandler().session_id['value']
+    yield mock_session_id
```
That is true, I will refactor this.

voigtst commented on tests/test_json_output.py:
```diff
@@ -0,0 +1,51 @@
+import yaml
+
+
+@pytest.fixture(scope='class')
```
That is true, I will refactor this.

voigtst replied:
done

voigtst commented:
> A more general note, consider using `self.assertEqual`, `self.assertFalse` etc to add error messages what is currently failing and why we expect that. For instance this is supposed to fail, because a login is missing: https://github.com/sapcc/vcsa-exporter/pull/7/files#diff-dfbf691ce15248c24632831234785fc221b10763ff99e326336d4ca328daa9d6R48
> which is not really obvious if you don't have insight.

I can't really mix pytest with unittest, because there are limitations for that. I wouldn't be able to  use non-autouse fixtures or parametrization for example. Refering to: https://docs.pytest.org/en/stable/unittest.html.
What I can do however is adding custom error messages which will be shown instead of the generic ones. I presume that's the same effect you are looking for. Please contact me to get more into detail with the desired requirements.

viennaa commented:
Very solid point. If you find a way, to make failing test reasons transparent, I would be pleased 👍 . At least for the asserts, where it is not really obvious. Got to add an inline comment.

viennaa commented on tests/test_json_output.py:
```diff
@@ -0,0 +1,27 @@
+        data = rest_yaml['logging']['action']
+        con = Connection(setup_vcenter)
+        response = con.post_request(api_target, data)
+        assert response is False
```
Here I would add some information why False is expected (missing login).

voigtst replied:
I implemented custom error messages.

voigtst commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+    pass
+
+
+def pytest_generate_tests(metafunc):
```
This is a very good point, my thought process here was to make it reusable throughout the tests, but it's clearly used once.
This issue correlates to the issues above. I wanted to check how the backend reacts to wrong credentials as well. Would it be sufficient to create one parameter with a wrong host and split up the 'test_login_logout_failure' test?

viennaa replied:
Well, for a login failure I would pass a wrong PW to have the MockServer denying this request and see how the code reacts.
Hence passing another host would result in a backend down scenario (which is also a valid thing to test, just not here).

voigtst replied:
Absolutely. Now I understand the requirement. Thanks, I will implement it this way.

voigtst replied:
I now splitted the 'test_login_logout_failure' to two seperate tests.
I also used the connection PW to cause a failed login.

voigtst commented on tests/conftest.py:
```diff
@@ -0,0 +1,78 @@
+
+@pytest.fixture(scope='session')
+def param():
+    """The param fixture allows us to pass all parameters from pytest_generate_tests to the target tests."""
```
This shouldn't be necessary anymore after adjusting the 'test_login_logout_failure' testcases.

voigtst replied:
I removed the parameters, which were used for one test, reducing the testcount from 12 to 9 (3 tests, 3 parameters).
I added the tests 'test_login_failure', 'test_logout_failure', increasing the testcount to 11.
I added the test 'test_vcenter_backend_up' which increases the testcount to 12 again.

Now we expect 4 Tests to fail, 8 to succeed, because 'test_vcenter_backend_up' succeeds.

voigtst commented on tests/test_connection.py:
```diff
@@ -0,0 +1,34 @@
+        setup_vcenter.login()
+        assert setup_vcenter.con.session_id == session_id
+
+    def test_login_logout_failure(self, param, host, mpw, user, pw):
```
done.

voigtst commented:
I removed the parameters, which were used for one test, reducing the testcount from 12 to 9 (3 tests, 3 parameters).
I added the tests 'test_login_failure', 'test_logout_failure', increasing the testcount to 11.
I added the test 'test_vcenter_backend_up' which increases the testcount to 12 again.

Now we expect 4 Tests to fail, 8 to succeed, because 'test_vcenter_backend_up' succeeds.

viennaa commented on tests/test_metric_output.py:
```diff
@@ -0,0 +1,50 @@
+        setup_vcenter.name = request.config.getoption('--host')
+        assert len(metrics) == 0, 'If the backend is down, no vcsa service metrics are expected'
+        # Backend up test
+        self.backend_up(setup_vcsa_url)
```
don't we need another assert, to ensure we are up again?

voigtst replied:
The assert is happening in the backend_up helper method for this testcase, reference to line 40 in this code. 'test_backend_recovery_from_down' calls 'beckend_up' which includes the assert. I can also return metrics and implement assert into this testcase. I think this would be more clear.
What do you think? :)

voigtst replied:
Changed it, I also changed the message at the second backend_up test.

viennaa reviewed (approved):
Let's do the GH integration, it is shouting for it. Thanks!
